### DISCUSSION TITLE
string is not stack allocated: not supported

### DISCUSSION THREAD

**Discussion by houminz** [cilium/ebpf#601] [Q&A] [answered]
Hi, I am trying to using `cilium/ebpf` to write ebpf program, and facing the problem of compiling.
The source code that I am writing is referred from [this blog](https://cyral.com/blog/how-to-ebpf-accelerating-cloud-native/) and [its github](https://github.com/cyralinc/os-eBPF/). I am porting it to Go, and here is my code [sockredir](https://github.com/SimpCosm/godemo/blob/master/sockredir).

The compiling message shows that `section "sockops": string is not stack allocated: not supported
`

``` $ make
go generate
Generating for amd64
Compiled /root/sockredir/sockredir_bpfel.o
Stripped /root/sockredir/sockredir_bpfel.o
Error: can't write /root/sockredir/sockredir_bpfel.go: can't load BPF from ELF: section "sockops": string is not stack allocated: not supported
exit status 1
main.go:18: running "go": exit status 1
make: *** [Makefile:15: BPFDemo] Error 1
```

<img width="1781" alt="image" src="https://user-images.githubusercontent.com/16812977/158836939-b3197bd1-c4eb-4bde-80da-dcce3b221c5f.png">

```main.go
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang SockRedir ./bpf/sockops.c -- -I./bpf/headers
```
The ELF object has been generated, but go file is not.

When I remove the code below, the build is successful. 
```
SEC("sockops")
int my_sockmap(struct bpf_sock_ops *skops)
{
    switch (skops->op) {
        case BPF_SOCK_OPS_PASSIVE_ESTABLISHED_CB:
        case BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB:
            if (skops->family == 2) { //AF_INET
                bpf_sock_ops_ipv4(skops);
            }
            break;
        default:
            break;
    }
    return 0;
}
```

I am wondering why shall we come across the error messages, since most of the code is brorrow from the repo and we could compile it using clang. 

Any help would be appreciated, thanks.

**Comment by chenhengqi** [comment, marked as answer]
Your problem is in `bpf_sock_ops_ipv4`: https://github.com/SimpCosm/godemo/blob/c3d3ab845732d415e4de1df8077c395daa7703f9/sockredir/bpf/sockops.c#L61-L63

Replace `bpf_trace_printk` with `bpf_printk`. The latter is a macro that ensures that the format string is allocated on the stack.
